### DISCUSSION TITLE
How to: Read data from cluster during a large import and index rebuild

### DISCUSSION THREAD

**Discussion by Stocki916** [ravendb/ravendb#16348] [Q&A]
Dear community,

we have a three node ravendb cluster with a large database which is replicated through all instances. Every night, we are updating multiple million documents (10 to 150 mio) and during the import the indezes are also updated.
But this index rebuild does take a long time to finish and for reading the data we ignore indizes with "stale" state, to be sure we are returning valid and consistent data. But this means, the application requests to database run in timeouts because the indexes are "stale" for a really long time.

In the "old" infrastructure (based on microsoft sql server) we used two seperate servers and used one server for application and updated the other (the offline instance). After import, we switched the servers.

Is there any idea or valueable approach you have in mind, to archive the goal to import data to the cluster AND read consistent data at the same time?
Maybe by importing and updating one node, and then (when all is done) start replication and use the finished node for reading?

**Comment by arekpalinski** [comment]
Hi,

If I understand correctly, during the import process you use RavenDB in read-only mode, is that correct? You could setup the external replication to a separate RavenDB replica instance from your cluster. Then before the import starts you disable the external replication task and switch the application to read from the offsite replica. Once the import completes and your indexes are up to date in your cluster then you switch the application back to it and turn on the external replication to propagate all updates to the replica.

https://ravendb.net/docs/article-page/5.4/csharp/server/ongoing-tasks/external-replication
https://ravendb.net/docs/article-page/5.4/csharp/studio/database/tasks/ongoing-tasks/external-replication-task

**Comment by ayende** [comment]
I assume that for most of the data, this is just an import from a source, most of which is not changed, right?
That being the case, we have `BulkInserOptions.SkipOverwriteIfUnchanged` that can *really* help in this scenario.

Only the changed documents will be actually updated in this case.

Beyond that, another option is to simple have two databases, and update each night the other one, then switch which one you are talking to.

How are you actually handling the updates? What kind of disk are you using?

**Comment by Stocki916** [comment]
No, the import just handles changed data. We already use the option "SkipOverwriteIfUnchanged" but the database uses a large amount of Autoindizes which take a lot of time to be rebuilded.
At the moment i prefer the idea with external replication: pause replication to one node, use this for reading and update the other nodes. After import, nodes will be switched and the "external" replication will be started.

By using two databases and switch it each night, we still have the problem that cluster performance is really bad (also with several timeout exceptions) during import.

We are using a SSD storage as disk below the nodes.

**Reply by garayx** [reply]
Are the source & replica databases on the same cluster ? if so you need to put the replica on different cluster, or make sure the node that has the replica wont serve the source database traffic.
